botocore/utils.py

```diff
@@ -602,6 +602,8 @@ def parse_timestamp(value):
     This will return a ``datetime.datetime`` object.
 
     """
+    # Prefer tzwinlocal when available to handle negative timestamps.
+    tzlocal = tzwinlocal or tzlocal
```

Comment:
Rather than changing our default to `tzwinlocal`, how do you feel about using it as a best effort fallback? That will preserve the current default behavior for most users, but enable graceful recovery.

The rough outline of what I'm thinking would be moving this functions content into a private function that takes the value and a new optional parameter `tzinfo`. From there we add some try/except logic to this function to first attempt parsing with `tzlocal` and then again with `tzwinlocal` if the `OSError` is raised.

e.g.
```python
def _parse_timestamp(value, tzinfo=tzlocal):
     #<current logic of parse_timestamp>

def parse_timestamp(value):
    tzinfo_options = (tzlocal, tzwinlocal)
    for tzinfo in tzinfo_options:
        try:
            return _parse_timestamp(value, tzinfo)
        except OSError as e:
            # debug logging about fallback
    # We shouldn't reach this point

Reply:
Hey @nateprewitt, thanks for taking a look. Your suggestion is even more gentle/conservative than the one I quickly typed up in my initial changeset while maintaining a similar spirit: find the best `tzinfo` value for the current environment. Nice work.

I will try out a test suite run with your changes out of curiosity to see if your changes produce fewer noise in CI with respect to test failures. If I recall correctly when I ran my changeset against some of the tests that failed on `botocore`'s Travis CI Linux containers they passed for me on a local `python:3.7-slim` (Debian) container. If your changes run without any noise in CI, all the better!

The bright side is your changeset has faith that the system may resolve successfully with `tzlocal` in some cases before ever needing to fall back to `tzwinlocal`, but the trade off is of course that if a Windows system remains in a problematic timezone, timestamps within the range that fail `tzlocal` will consistently require a second attempt (and will log to debug or warn level accordingly). But only certain ranges of timestamps fail depending on the timezone, so perhaps it is fair to mention your change will not directly double the total number of attempts.

My original suggestion relied on a different angle of optimism: on Windows, regardless of the timezone, `tzwinlocal` (so far) provides us a `tzinfo` object that will yield the correct result for all timestamps in the underlying `time.localtime(...)` call.

There's no response yet upstream from the `dateutil` authors as to what they plan to do with respect to `tz.gettz()` for issue [#848](https://github.com/dateutil/dateutil/issues/848) beyond acknowledging that this is a known portability issue, but regardless of how that issue is resolved a change along the lines of what we have proposed here in `botocore` will be required to ensure the this part of the `botocore` library works on Windows as well as it does on POSIX systems.

Reply:
Yep, I think simplifying to just using `tz.gettz()` is the long term goal if #848 gets resolved. We can probably make a note that as a todo.

I agree this is a slight performance hit with the retry and may be some additional noise. The thought process for the tradeoff is it's only for the edge case, and it avoids having to do more extensive testing around things like bad registry entries and any other potential anomalies on windows. We might look at a log filter at a later date if we find the debug noise is excessive.